### PR TITLE
Prototype of ASDF

### PR DESCRIPTION
**Pull request by OrDTesters**

### CODE REVIEW

File: dt-impact-tracer/dt-impact-tracer.iml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
Comment (ReedOei) [inline dt-impact-tracer/dt-impact-tracer.iml:1]: This isn't really related to this pull request specifically, but I think we should probably stop tracking these IDE files. They add a lot of clutter to diffs and don't seem to really provide any benefit.

File: dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/Main/AutoCleanerMain.java [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        this.testBinaryPath = Paths.get(getArgRequired("testBinaryPath"));
+    }
+
+    private boolean didTestsPass(List<String> tests) throws IOException {
+        FailingTestDetector detector = getFailingTestDetector(classpath, tests);
+        Set<String> failedTest = detector.notPassingTests();
+        return failedTest.isEmpty();
+    }
+
+    private FailingTestDetector detector;
```
Comment (ReedOei) [inline dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/Main/AutoCleanerMain.java]: After making other changes I suggested, we can initialize this in the constructor/make it final, rather than having the `getFailingTestDetector`.

File: dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/Main/AutoCleanerMain.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+            if (jFile.getMethodDeclaration(testName) != null) {
+                return jFile;
+            }
+        }
+        return null;
+    }
+
+    private void saveOrigFile(Path filePath) throws IOException {
+        Path parent = filePath.getParent();
+        Path copyFilePath = Paths.get(parent.toString(), filePath.getFileName().toString() + ".orig");
```
Comment (ReedOei) [inline dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/Main/AutoCleanerMain.java]: Could use `copyFilePath = parent.resolve(filePath.getFileName().toString() + ".orig");` instead of converting everything to strings.

File: dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/detectors/FailingTestDetector.java [hunk 4]
```diff
@@ -57,24 +60,38 @@ private void output(final Set<String> tests) throws IOException {
...
 
-    private Set<String> notPassingTests() throws IOException {
-        final List<String> tests = Files.readAllLines(testList, Charset.defaultCharset());
+    public FailingTestDetector(String classpath, List<String> tests) {
+        super(new String[0]);
+        this.classpath = classpath;
+        this.tests = tests;
+    }
+
+    public void setTestList(List<String> tests) {
```
Comment (ReedOei) [inline dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/detectors/FailingTestDetector.java]: Would probably be better to pass `tests` to `notPassingTests()` as an argument rather than separately setting it. This way it can also be removed as a field altogether.

File: dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/detectors/FailingTestDetector.java [hunk 5]
```diff
@@ -57,24 +60,38 @@ private void output(final Set<String> tests) throws IOException {
...
             notPassingTests.addAll(tests.stream().filter(t -> !result.getResult(t).result.equals(RESULT.PASS)).collect(Collectors.toList()));
+
             tests.removeAll(notPassingTests);
             final int afterSize = tests.size();
 
             if (beforeSize == afterSize) {
                 break;
             }
         }
+        System.out.println();
```
Comment (ReedOei) [inline dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/detectors/FailingTestDetector.java]: Why is this here? This method doesn't seem to do any other printing.

File: dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/detectors/FailingTestDetector.java [hunk 6]
```diff
@@ -57,24 +60,38 @@ private void output(final Set<String> tests) throws IOException {
...
+    public void setTestList(List<String> tests) {
+        this.tests = tests;
+    }
+
+    public Set<String> notPassingTests() throws IOException {
         final Set<String> notPassingTests = new HashSet<>();
 
         ImpactMain.skipMissingTests = true;
 
         while (true) {
```
Comment (ReedOei) [inline dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/detectors/FailingTestDetector.java]: It looks sort of weird to have this if inside the while loop. Maybe we could just change the condition of the while loop to be `!tests.isEmpty()`, or return early if the test list passed in is empty.

File: dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/data/JavaFile.java [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+            }
+        }
+
+        return null;
+    }
+
+    private String getFullyQualifiedMethodName(MethodDeclaration method, ClassOrInterfaceDeclaration classDec) {
+        String packageName = compilationUnit.getPackageDeclaration().get().getName().toString();
+        String className = classDec.getNameAsString();
+        String fullMethodName = String.format("%s.%s.%s", packageName, className, method.getName().getIdentifier());
```
Comment (ReedOei) [inline dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/data/JavaFile.java]: Creating `fullMethodName` is unnecessary, we can just directly return the value.

File: dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/data/JavaFile.java [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+
+    public String getFilename() {
+        return filename;
+    }
+
+    /**
+     * Finds all classes/interfaces in the file and saves them.
+     * @param outfileName
+     */
+    public void loadClassList(final Path outfileName) throws IOException {
```
Comment (ReedOei) [inline dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/data/JavaFile.java]: It's weird that this method exists, as nothing will work until this method is called. It's logic should probably be part of the constructor or some factory method.

Also, having the `outfileName` as a parameter is weird because it's passed into the constructor.

File: dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/Main/AutoCleanerMain.java [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+        this.classpath = FailedTestRemover.buildClassPath(getArgRequired("classpath").split(":"));
+        this.cleanerTestName = getArgRequired("cleanerTestName");
+        this.victimTestName = getArgRequired("victimTestName");
+        this.testFilesPath = Paths.get(getArgRequired("tests"));
+        this.testBinaryPath = Paths.get(getArgRequired("testBinaryPath"));
+    }
+
+    private boolean didTestsPass(List<String> tests) throws IOException {
+        FailingTestDetector detector = getFailingTestDetector(classpath, tests);
+        Set<String> failedTest = detector.notPassingTests();
```
Comment (ReedOei) [inline dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/Main/AutoCleanerMain.java]: Useless local variable `failedTest`

File: dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/Main/AutoCleanerMain.java [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+            detector = new FailingTestDetector(classpath, new ArrayList<>(tests));
+        } else {
+            detector.setTestList(new ArrayList<>(tests));
+        }
+        return detector;
+    }
+
+    private void compile(final JavaFile javaFile) throws Exception {
+        int errorCount = -1;
+        while (errorCount != 0) {
```
Comment (ReedOei) [inline dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/Main/AutoCleanerMain.java]: This loop doesn't seem to do anything, as `errorCount` is directly set to 0 at the end of the loop, and is not actually used at any point. This also leads to the method printing "Successfully compiled" regardless of whether it was actually successfully compiled. The method should probably be modified to at least return whether compilation was successful.

File: dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/Main/AutoCleanerMain.java [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+                                       + " with " + errorCount + " errors.");
+            errorCount = 0;
+        }
+
+        System.out.println("Successfully compiled at: " + new Date().toString());
+        System.out.println();
+    }
+
+    private JavaFile getJavaFile(String testName, List<Path> testFiles, boolean saveOrigFile, Path outputPath) throws IOException {
+        // For each test source files, find the cleaner and victim test name
```
Comment (ReedOei) [inline dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/Main/AutoCleanerMain.java]: This comment seems wrong. The method looks like it finds the `JavaFile` containing the `testName` passed in, and does not "find the...test name", and has no specific relation to the cleaner or victim tests.

File: dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/Main/AutoCleanerMain.java [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+            Path pOrig = Paths.get(p.toString() + ".orig");
+
+            if (Files.exists(pOrig)) {
+                Files.copy(pOrig, p,
+                           StandardCopyOption.REPLACE_EXISTING);
+            } else if (saveOrigFile){
+                saveOrigFile(p);
+            }
+
+            // Copy the compiled binary to the correct location
```
Comment (ReedOei) [inline dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/Main/AutoCleanerMain.java]: The code below seems to check whether a file contains the `testName`, not copy binaries?

File: dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/Main/AutoCleanerMain.java [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+        }
+
+        JavaFile cleanerJavaFile = getJavaFile(cleanerTestName, testFiles, false, testBinaryPath);
+        MethodDeclaration cleanerMethod = cleanerJavaFile.getMethodDeclaration(cleanerTestName);
+        BlockStmt cleanerBlock = cleanerMethod.getBody().get();
+
+        JavaFile victimJavaFile = getJavaFile(victimTestName, testFiles, true, testBinaryPath);
+
+        // Check if we pass in isolation before fix
+        System.out.println("[INFO] Found and compiling victim test.");
```
Comment (ReedOei) [inline dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/Main/AutoCleanerMain.java:128]: Why is this check here? It doesn't seem like the result from `didTestsPass()` is actually used.

File: dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/Main/README.md [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+  -victimTestName org.apache.zeppelin.dep.DependencyResolverTest.testDelRepo \ 
+  -classpath $NEW_DT_TESTS:$NEW_DT_CLASS:$NEW_DT_LIBS \
+  -testBinaryPath /Users/winglam/tmp/zeppelin/zeppelin-interpreter/target/test-classes/org/apache/zeppelin/dep/DependencyResolverTest.java
+```
+
++ ```-tests```: directory containing the source files that contains the cleaner and victim test methods.
++ ```-cleanerTestName```: fully qualified method name of the cleaner.
++ ```-victimTestName```: fully qualified method name of the victim.
++ ```-classpath```: classpath of compiled source and test code of the subject and the libs that the subject depends on. Libs can be expressed with `*` as the program will expand such wildcards.
++ ```-testBinaryPath```: directory containing the compiled victim test.
```
Comment (ReedOei) [inline dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/Main/README.md]: Looks like the actual fixed source file (or some source file, at least), rather than a directory or a binary file.

File: dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/Main/AutoCleanerMain.java [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+            // TODO a better way should be implemented to look up the location of the
+            // binary file using the source file that we found, fixed, and recompiled
+            JavaFile jFile = new JavaFile(p.getFileName().toString(), classpath, outputPath);
+
+            jFile.loadClassList(p);
+            if (jFile.getMethodDeclaration(testName) != null) {
+                return jFile;
+            }
+        }
+        return null;
```
Comment (ReedOei) [inline dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/Main/AutoCleanerMain.java]: This method returns null, but all methods that call it assume it doesn't. Would be good to add some error checking at some point.

**Review by ReedOei — COMMENTED** [review]
@winglam I was able to get the tool to work using your instructions. I finished my review; most of the changes I suggested are fairly minor--not sure if you were planning to fix anyway.

With regards to the improvements to make this tool what we discussed (e.g., copying before/after/delta-debugging) is the plan for me to implement that when I merge this into the dt-fixing-tools?

File: dt-impact-tracer/dt-impact-tracer.iml (hunk 1 above)
Comment (OrDTesters) [reply dt-impact-tracer/dt-impact-tracer.iml:1]: There are lots of posts online discussing whether IDE specific files should or should not be VC'd. My thoughts on the matter as of now would be to allow them still. If they later cause many unnecessary merge conflicts, we'll omit these files then. 
https://intellij-support.jetbrains.com/hc/en-us/community/posts/206838875-IntelliJ-files-iml-ipr-etc-and-Source-Control?page=1#community_comment_206318055

File: dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/Main/AutoCleanerMain.java (hunk 12 above)
Comment (OrDTesters) [reply dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/Main/AutoCleanerMain.java]: loadClassList will compile the binary to the location of outputPath. I have clarified the comment. Thanks.

File: dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/data/JavaFile.java (hunk 8 above)
Comment (OrDTesters) [reply dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/data/JavaFile.java]: loadClassList does need to be able to take a parameter because most of the time it would load the outfilename but there is a time where we actually set outfilename and pass in something different as the parameter. The use case here is along the lines of parsing one file but outputting to a different file.

For clarity I have temporarily changed the name of the outfileName in this method to be different than the class-level outfileName variable. We can look to separate the use case into two variables at a later time.

File: dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/detectors/FailingTestDetector.java (hunk 5 above)
Comment (OrDTesters) [reply dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/detectors/FailingTestDetector.java]: Largely because FixedOrderRunner outputs a bunch of stuff and it clutters the output without the line break afterward. I have moved this to the AutoCleanerMain for now though since I guess not everyone who may use the FailingTestDetector would want the line break.

**Comment by OrDTesters** [comment]
> With regards to the improvements to make this tool what we discussed (e.g., copying before/after/delta-debugging) is the plan for me to implement that when I merge this into the dt-fixing-tools?

You don't have to implement the delta debugging. Just move the tool over to the dt-fixing-tools repository and reduce any unnecessary code as you see fit (e.g., use of FailingTestDetector vs. testrunner). 

Once you move the code over we can decide among the three of us who will continue on the next steps of this tool.

Thanks for all of the feedback. They were quite useful!
